Epona/include/Epona/Debug.hpp

```diff
@@ -0,0 +1,60 @@
+/*
+* Copyright (C) 2017 by Godlike
+* This code is licensed under the MIT license (MIT)
+* (http://opensource.org/licenses/MIT)
+*/
+#ifndef EPONA_DEBUG_HPP
+#define EPONA_DEBUG_HPP
+
+#include <functional>
+#include <vector>
+#include <glm/glm.hpp>
+
+namespace epona {
+    template <typename T>
+    class QuickhullConvexHull;
+} // namespace epona
+
+namespace epona
+{
+namespace debug
+{
+
+/* Singleton class that provides callback interface for debugging */
+class Debug
+{
+public:
+    Debug() = default;
+    Debug& operator==(Debug const&) = delete;
+    Debug(Debug&) = delete;
+    Debug& operator==(Debug&&) = delete;
+    Debug(Debug&&) = delete;
+
+    /* Returns reference to the singleton instance of the class */
+    static Debug& GetInstace();
+
+    /**
+     * @brief QuickhullConvexHull debug call function
+     *
+     * This method is called from within the QuickhullConvexHull class
+     * during the calculation. It than proxies all the arguments to the currently
+     * set callback functor.
+     *
+     * @param[in] convexHull   current convex hull instance
+     * @param[in] vertexBuffer source vertices for the convex hull
+     */
+    static void QuickhullConvexHullCall(
+        QuickhullConvexHull<std::vector<glm::dvec3>>& convexHull,
+        std::vector<glm::dvec3>& vertexBuffer
+    );
```

Comment:
Technically you could use `#ifdef` here to improve optimization by compiler and make it easier to follow:

```cpp
static void QuickhullConvexHullCall(
    QuickhullConvexHull<std::vector<glm::dvec3>>& convexHull,
    std::vector<glm::dvec3>& vertexBuffer
)
#ifdef EPONA_DEBUG
;
#else
{}
#endif
```

Another alternative would be to use 2 specific classes that share an interface but one has only dummy method bodies (`{}`) and the other would have the actual implementation for debug interfaces. This file would then just be
```cpp

#ifdef EPONA_DEBUG
#include <epona/debug/DebugImplementation.hpp>
#else
#include <epona/debug/DebugDummy.hpp>
#endif

namespace epona
{
namespace debug
{
#ifdef EPONA_DEBUG
using Debug = DebugImplementation;
#else
using Debug = DebugDummy;
#endif
}
}
```

Reply:
I think the Dummy one involves more code, therefore, the first one is better.

Reply:
Right now yes, but if the interface gets more busy it's cleaner to have a specific dummy substitute instead of a bunch of `#ifdef` conditionals for every method